components/resizable/nz-resizable.directive.ts

```diff
@@ -0,0 +1,325 @@
+        this.endResize($event);
+      }
+    });
+  }
+
+  @HostListener('document:mousemove', ['$event'])
+  onMousemove($event: MouseEvent): void {
+    this.ngZone.runOutsideAngular(() => {
+      if (this.resizing) {
+        this.resize($event);
+      }
+    });
+  }
+
+  onMouseenter(): void {
+    this.nzResizableService.mouseEntered$.next(true);
+  }
+
+  onMouseleave(): void {
+    this.nzResizableService.mouseEntered$.next(false);
+  }
+
+  setPosition(): void {
+    const position = getComputedStyle(this.el).position;
+    if (position === 'static' || !position) {
+      this.renderer.setStyle(this.el, 'position', 'relative');
+    }
+  }
+
+  calcSize(width: number, height: number, ratio: number): NzResizeEvent {
+    let newWidth: number;
+    let newHeight: number;
+    let col = 0;
+    let spanWidth = 0;
+    let maxWidth = Infinity;
+    let minWidth = this.nzMinWidth;
+    let maxHeight = Infinity;
+    if (this.nzBounds === 'parent') {
+      const parent = this.renderer.parentNode(this.el);
+      if (parent instanceof HTMLElement) {
+        const parentRect = parent.getBoundingClientRect();
+        maxWidth = this.nzMaxWidth
+          ? this.nzMaxWidth < parentRect.width
+            ? this.nzMaxWidth
+            : parentRect.width
+          : parentRect.width;
+        maxHeight = this.nzMaxHeight
+          ? this.nzMaxHeight < parentRect.height
+            ? this.nzMaxHeight
+            : parentRect.height
+          : parentRect.height;
+      }
+    } else if (this.nzBounds === 'window') {
+      if (typeof window !== 'undefined') {
+        const windowWidth = window.innerWidth;
+        const windowHeight = window.innerHeight;
+        maxWidth = this.nzMaxWidth ? (this.nzMaxWidth < windowWidth ? this.nzMaxWidth : windowWidth) : windowWidth;
+        maxHeight = this.nzMaxHeight
+          ? this.nzMaxHeight < windowHeight
+            ? this.nzMaxHeight
+            : windowHeight
+          : windowHeight;
+      }
+    } else if (this.nzBounds && this.nzBounds.nativeElement && this.nzBounds.nativeElement instanceof HTMLElement) {
+      const boundsRect = this.nzBounds.nativeElement.getBoundingClientRect();
+      maxWidth = this.nzMaxWidth
+        ? this.nzMaxWidth < boundsRect.width
+          ? this.nzMaxWidth
+          : boundsRect.width
+        : boundsRect.width;
+      maxHeight = this.nzMaxHeight
+        ? this.nzMaxHeight < boundsRect.height
+          ? this.nzMaxHeight
+          : boundsRect.height
+        : boundsRect.height;
+    }
+
+    if (this.nzGridColumnCount !== -1) {
+      spanWidth = maxWidth / this.nzGridColumnCount;
+      minWidth = this.nzMinColumn !== -1 ? spanWidth * this.nzMinColumn : minWidth;
+      maxWidth = this.nzMaxColumn !== -1 ? spanWidth * this.nzMaxColumn : maxWidth;
+    }
+
+    if (ratio !== -1) {
+      if (/(left|right)/i.test(this.currentHandleEvent.direction)) {
+        newWidth = Math.min(Math.max(width, minWidth), maxWidth);
```

Comment:
The same, it could be a function like `ensureInBoundary`.

It occurred to me that I wrote a function like this in the carousel component. You can extract that to `core`.